piker/fsp/_volume.py

```diff
@@ -174,4 +202,132 @@ async def dolla_vlm(
+
+    # TODO (idea): a dynamic generic / boxing type that can be updated by other
+    # FSPs, user input, and possibly any general event stream in
+    # real-time. Hint: ideally implemented with caching until mutated
+    # ;)
+    period: 'Param[int]' = 16,  # noqa
+
+    # TODO (idea): a generic for declaring boxed fsps much like ``pytest``
+    # fixtures? This probably needs a lot of thought if we want to offer
+    # a higher level composition syntax eventually (oh right gotta make
+    # an issue for that).
+    # ideas for how to allow composition / intercalling:
+    # - offer a `Fsp.get_history()` to do the first yield output?
+    #  * err wait can we just have shm access directly?
+    # - how would it work if some consumer fsp wanted to dynamically
+    # change params which are input to the callee fsp? i guess we could
+    # lazy copy in that case?
+    # dvlm: 'Fsp[dolla_vlm]'
+
+) -> AsyncIterator[
+    tuple[str, Union[np.ndarray, float]],
+]:
+    # generally no history available prior to real-time calcs
+    yield {
+        # from ib
+        '1m_trade_rate': None,
+        '1m_vlm_rate': None,
+
+        'trade_rate': None,
+        'dark_trade_rate': None,
+
+        'dvlm_rate': None,
+        'dark_dvlm_rate': None,
+    }
+
+    ltr = 0
+    lvr = 0
+
+    # TODO: 3.10 do ``anext()``
+    quote = await source.__anext__()
+    tr = quote.get('tradeRate')
+    yield '1m_trade_rate', tr or 0
+    vr = quote.get('volumeRate')
+    yield '1m_vlm_rate', vr or 0
+
+    # NOTE: in theory we could dynamically allocate a cascade based on
+    # this call but not sure if that's too "dynamic" in terms of
+    # validating cascade flows from message typing perspective.
+
+    # attach to ``dolla_vlm`` fsp running
+    # on this same source flow.
+    dvlm_shm = dolla_vlm.get_shm(ohlcv)
+
+    # precompute arithmetic mean weights (all ones)
+    seq = np.full((period,), 1)
+    weights = seq / seq.sum()
+
+    async for quote in source:
+        if not quote:
+            log.error("OH WTF NO QUOTE IN FSP")
+            continue
+
+        dvlm_wma = _wma(
+            dvlm_shm.array['dolla_vlm'],
+            period,
+            weights=weights,
+        )
+        yield 'dvlm_rate', dvlm_wma[-1]
+        trade_rate_wma = _wma(
+            dvlm_shm.array['trade_count'],
+            period,
+            weights=weights,
+        )
+        yield 'trade_rate', trade_rate_wma[-1]
+
+        # TODO: skip this if no dark vlm is declared
+        # by symbol info (eg. in crypto$)
+        dark_dvlm_wma = _wma(
+            dvlm_shm.array['dark_vlm'],
+            period,
+            weights=weights,
+        )
+        yield 'dark_dvlm_rate', dark_dvlm_wma[-1]
+
+        dark_trade_rate_wma = _wma(
+            dvlm_shm.array['dark_trade_count'],
+            period,
+            weights=weights,
+        )
+        yield 'dark_trade_rate', dark_trade_rate_wma[-1]
+
+        # XXX: ib specific schema we should
```

Comment:
This is commented out rates from ib that i'm not sure we'll even want since as mentioned they're only updated every 25s..